jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -1456,6 +1456,13 @@ private DynamicCapacity(List<RetainableByteBuffer> buffers, ByteBufferPool.Sized
                 throw new IllegalArgumentException("must always retain if cannot aggregate");
         }
 
+        private void checkNotReleased()
+        {
+            ReferenceCounter counter = (ReferenceCounter)getWrapped();
+            if (counter.get() == 0)
+                throw new IllegalStateException("Already released");
```

Comment:
Use the API:
```suggestion
            if (getRetained() <= 0)
                throw new IllegalStateException("Already released");
```

Reply:
Actually I'd prefer:
```suggestion
            assert getRetained() > 0;
```

and then inline the method.

Reply:
or maybe:
```suggestion
            if(LOG.isDebugEnabled() && getRetained() <= 0)
                throw new IllegalStateException("Already released");
```

Reply:
I changed the code to use the API as suggested, but I'm not in favor of only enforcing this check when assertions or debug logs are enabled as IMHO the atomic get is cheap enough that it's virtually free in practice.